internal/scheduler/scheduling/optimiser/node_scheduler.go

```diff
@@ -0,0 +1,285 @@
+			}
+			if jctx.PodSchedulingContext == nil {
+				return nil, fmt.Errorf("no pod scheduling context exists on jctx for job %s despite it being in successful scheduling contexts", jobId)
+			}
+			scheduledAtPriority = jctx.PodSchedulingContext.ScheduledAtPriority
+		} else {
+			if job.LatestRun() == nil {
+				return nil, fmt.Errorf("no job run found for scheduled job %s", jobId)
+			}
+			if job.LatestRun().ScheduledAtPriority() == nil {
+				return nil, fmt.Errorf("scheduled at priority is nil for scheduled job %s", jobId)
+			}
+			scheduledAtPriority = *job.LatestRun().ScheduledAtPriority()
+			age = start.Sub(*job.LatestRun().LeaseTime()).Milliseconds()
+		}
+		if scheduledAtPriority > jobToSchedule.Job.PriorityClass().Priority {
+			// Can't evict jobs of higher priority
+			continue
+		}
+
+		cost := schedContext.Sctx.FairnessCostProvider.UnweightedCostFromAllocation(resource)
+		runInfo := &preemptibleJobDetails{
+			cost:                cost,
+			resources:           resource,
+			jobId:               jobId,
+			queue:               queue,
+			scheduledAtPriority: scheduledAtPriority,
+			ageMillis:           age,
+		}
+
+		if _, present := queues[queue]; !present {
+			queues[queue] = []*preemptibleJobDetails{runInfo}
+		} else {
+			queues[queue] = append(queues[queue], runInfo)
+		}
+	}
+	return queues, nil
+}
+
+func (n *PreemptingNodeScheduler) populateQueueImpactFields(schedContext *SchedulingContext, jobToSchedule *context.JobSchedulingContext, queues map[string][]*preemptibleJobDetails) error {
+	for queue, items := range queues {
+		sort.Sort(internalQueueOrder(items))
+
+		qctx, ok := schedContext.Queues[queue]
+		if !ok {
+			return fmt.Errorf("queue context for queue %s not found", queue)
+		}
+
+		updatedQueueCost := qctx.CurrentCost
+		count := 0
+		for _, item := range items {
+			updatedQueueCost = roundFloatHighPrecision(updatedQueueCost - item.cost)
+			item.weightedCostAfterPreemption = updatedQueueCost / qctx.Weight
+			if item.scheduledAtPriority < jobToSchedule.Job.PriorityClass().Priority {
+				item.costToPreempt = 0
+				item.priorityPreemption = true
+			} else if updatedQueueCost > qctx.Fairshare {
+				item.costToPreempt = 0
+			} else {
+				// This could be improved to handle crossing the fairshare boundary better
+				// It could just cost the amount the queue is brought below fairshare, rather than the full item cost
+				item.costToPreempt = item.cost
+			}
+			item.queuePreemptedOrdinal = count
+			count++
+		}
+
+		queues[queue] = items
+	}
+	return nil
+}
+
+func roundFloatHighPrecision(input float64) float64 {
+	return math.Round(input*100000000) / 100000000
+}
+
+func isTooLargeToEvict(job *jobdb.Job, limit *internaltypes.ResourceList) bool {
+	if limit == nil {
+		return false
+	}
+	if job == nil {
+		return true
+	}
+	jobResources := job.AllResourceRequirements()
+	if jobResources.Factory() != limit.Factory() {
+		return true
```

Comment:
There should only be a single global factory (except when unit testing).
Panic here rather than returning true?